src/tribler/core/utilities/search_utils.py

```diff
@@ -27,3 +32,194 @@ def split_into_keywords(string, to_filter_stopwords=False):
+
+
+word_re = re.compile(r'\w+', re.UNICODE)
+
+
+def title_rank(query: str, title: str) -> float:
+    """
+    Calculate the similarity of the title string to a query string, with or without stemming.
+    """
+    query = word_re.findall(query.lower())
+    title = word_re.findall(title.lower())
+    return calculate_rank(query, title)
+
+
+# These coefficients are found empirically. Their exact values are not very important for a relative ranking of results
+
+# The first word in a query is considered as a more important than the next one and so on,
+# 5 means the 5th word in a query is twice as less important as the first one
+POSITION_COEFF = 5
+
+# Some big value for a penalty if a query word is totally missed from a torrent title
+MISSED_WORD_PENALTY = 10
+
+# If a torrent title contains some words at the very end that are not mentioned in a query, we add a very slight
+# penalty for them. The *bigger* the REMAINDER_COEFF is, the *smaller* penalty we add for this excess words
+REMAINDER_COEFF = 10
+
+# The exact value of this coefficient is not important. It is used to convert total_error value to a rank value.
+# The total_error value is some positive number. We want to have the resulted rank in range [0, 1].
+RANK_NORMALIZATION_COEFF = 10
+
+
+def calculate_rank(query: List[str], title: List[str]) -> float:
+    """
+    Calculate the similarity of the title to the query as a float value in range [0, 1].
+    """
+    if not query:
+        return 1.0
+
+    if not title:
+        return 0.0
+
+    title = deque(title)
+    total_error = 0
+    for i, term in enumerate(query):
+        # The first word is more important than the second word, and so on
+        term_weight = POSITION_COEFF / (POSITION_COEFF + i)
+
+        # Read the description of the `find_term` function to understand what is going on. Basically, we are trying
+        # to find each query word in the title words, calculate the penalty if the query word is not found or if there
+        # are some title words before it, and then rotate the skipped title words to the end of the title. This way,
+        # the least penalty got a title that has query words in the proper order at the beginning of the title.
+        found, skipped = find_term(term, title)
+        if found:
+            # if the query word is found in the title, add penalty for skipped words in title before it
+            total_error += skipped * term_weight
+        else:
+            # if the query word is not found in the title, add a big penalty for it
+            total_error += MISSED_WORD_PENALTY * term_weight
+
+    # a small penalty for excess words in the title that was not mentioned in the search phrase
+    remainder_weight = 1 / (REMAINDER_COEFF + len(query))
+    remained_words_error = len(title) * remainder_weight
+    total_error += remained_words_error
+
+    # a search rank should be between 1 and 0
+    return RANK_NORMALIZATION_COEFF / (RANK_NORMALIZATION_COEFF + total_error)
```

Comment:
I am wondering... could `calculate_rank` be replaced by something simpler?

After a quick research, I prepared a test function that satisfies @synctext requirements posted [here](https://github.com/Tribler/tribler/issues/2250#issuecomment-1157735819) 

```python
def test_calculate_rank():
    # rule 2
    assert torrent_rank(query='Sintel', title='Sintel') > \
           torrent_rank(query='Sintel', title='the.script.from.the.movie.Sintel.pdf')

    # rule 3
    assert torrent_rank(query='Sintel', title='Sintel-Part-I') > \
           torrent_rank(query='Sintel', title='Part-of-Sintel')

    # rule 4
    assert torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy") > \
           torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy: The Story") >\
           torrent_rank(query="The Internet's Own Boy", title="The Internet's Own Boy: The Story of Aaron Swartz")

    # rule 5
    assert torrent_rank(query="Internet's Own Boy", title="Internet's Own Boy - Aaron Swartz") > \
           torrent_rank(query="Internet's Own Boy", title="Internet's very Own Boy")

    assert torrent_rank(query="Internet's Own Boy", title="Internet's Own Boy - Aaron Swartz") > \
           torrent_rank(query="Internet's Own Boy", title="Internet's very special Boy person")
```

And came up with the simple `calculate_rank` function that is probably good enough for the first iteration of search improvement.

The new version:
```python
def calculate_rank(query: List[str], title: List[str]) -> float:
    title = ' '.join(title)
    query = ' '.join(query)
    matcher = SequenceMatcher(None, query, title, autojunk=False)
    longest = matcher.find_longest_match(0, len(query), 0, len(title))
    return matcher.quick_ratio() * longest.size
```

The old version:
```python
def calculate_rank(query: List[str], title: List[str]) -> float:
    if not query:
        return 1.0

    if not title:
        return 0.0

    title = deque(title)
    total_error = 0
    for i, term in enumerate(query):
        # The first word is more important than the second word, and so on
        term_weight = POSITION_COEFF / (POSITION_COEFF + i)

        # Read the description of the `find_term` function to understand what is going on. Basically, we are trying
        # to find each query word in the title words, calculate the penalty if the query word is not found or if there
        # are some title words before it, and then rotate the skipped title words to the end of the title. This way,
        # the least penalty got a title that has query words in the proper order at the beginning of the title.
        found, skipped = find_term(term, title)
        if found:
            # if the query word is found in the title, add penalty for skipped words in title before it
            total_error += skipped * term_weight
        else:
            # if the query word is not found in the title, add a big penalty for it
            total_error += MISSED_WORD_PENALTY * term_weight

    # a small penalty for excess words in the title that was not mentioned in the search phrase
    remainder_weight = 1 / (REMAINDER_COEFF + len(query))
    remained_words_error = len(title) * remainder_weight
    total_error += remained_words_error

    # a search rank should be between 1 and 0
    return RANK_NORMALIZATION_COEFF / (RANK_NORMALIZATION_COEFF + total_error)


def find_term(term: str, title: Deque[str]) -> Tuple[bool, int]:
    try:
        skipped = title.index(term)
    except ValueError:
        return False, 0

    title.rotate(-skipped)  # rotate skipped words to the end
    title.popleft()  # remove found word
    return True, skipped
```

Reply:
If I got you correctly, you have the following concerns about the suggested function:
1. The output is not normalized to [0..1]
2. It works worse for one corner case: "the excess words at the beginning and the end of the title".

About the (1) -- yes, it is a draft. The output could be easily normalized. I didn't do it because for the prototype it is not necessary.

About (2) yes, but the suggested solution works pretty well for the most cases and extremely easy in terms of implementation.

Pros and cons in the single table:
|   | Original Solution | Suggested Solution |
| ------------- | ------------- |------------- |
| Johan Test  | ✅ | ✅ |
| Speed  | 1.4 | 20.8 |
| LOC  | 100 | 6 |
| Kozlovsky Test (corner cases)  | ✅ | ❌ |

The speed test has been performed by the similar test that you did before:

```python
import timeit
from difflib import SequenceMatcher
from typing import List


def calculate_rank2(query: List[str], title: List[str]) -> float:
    title = ' '.join(title)
    query = ' '.join(query)
    matcher = SequenceMatcher(None, query, title, autojunk=False)
    longest = matcher.find_longest_match(0, len(query), 0, len(title))
    return matcher.quick_ratio() * longest.size


t1 = timeit.Timer(
    "calculate_rank('The Big Buck Bunny', 'Making of The Big Buck Bunny')",
    "from tribler.core.utilities.search_utils import calculate_rank\n")

t2 = timeit.Timer(
    "calculate_rank2('The Big Buck Bunny', 'Making of The Big Buck Bunny')",
    "from __main__ import calculate_rank2\n")

print('For the test, time 10k invocation of function, repeat 5 times')

function1_results = t1.repeat(repeat=5, number=10000)
print(f'Original function speed: {function1_results}')

function2_results = t2.repeat(repeat=5, number=10000)
print(f'Suggested function speed: {function2_results}')

# Use the fastest result from each benchmark as it is the least affected by other parallel tasks and so the most precise
slowdown = min(function2_results) * 100 / min(function1_results) - 100
print(f'The slowdown is {slowdown} percent')
```

To summarize the above my opinion is that we should keep your solution despite its complexity and size 🤝

Reply:
Actually, the performance test requires a small fix. On my machine, the test shows that the function based on `SequenceMatcher` is 1579% slower than the solution from the PR. But the `calculate_rank` function actually expects lists of words, not strings, in its arguments. So, the correct timers definitions should look like this:

```python
t1 = timeit.Timer(
    "calculate_rank('The Big Buck Bunny'.split(), 'Making of The Big Buck Bunny'.split())",
    "from tribler.core.utilities.search_utils import calculate_rank\n")

t2 = timeit.Timer(
    "calculate_rank2('The Big Buck Bunny'.split(), 'Making of The Big Buck Bunny.split()')",
    "from __main__ import calculate_rank2\n")
```

With this change, the function based on SequenceMatcher is still slower, but only 849% slower.

Reply:
> the suggested solution works pretty well for the most cases

Actually, I consider the remaining cases quite significant and not so rare. It is pretty possible that Tribler has an almost perfect title match in the database, but it receives a heavy penalty for additional words at the end of the title name and, as a result, is not included in the search result list. With the function implementation from the PR, such results will not be accidentally penalized, and users will see them. I expect it to be important for around 10% of search queries.

Reply:
In other words, if a user searches for "The Big Buck Bunny 1080p", and the title in the database is "The Big Buck Bunny MP4 1080p 30FPS Eng", it should almost not be penalized for a non-perfect match, and the algorithm from the PR takes that into account.